### PR TITLE
Document exceptions in callbacks

### LINKED ISSUE #5434: Document exceptions
## 📚 Documentation
This is only a suggestion, but I find it useful to have what exceptions functions/classes can raise in the docs, so I'd like to suggest adding the `Raises:` section to public functions/classes.

**Progress**
`dirname $(grep -rl "raise " pytorch_lightning)|sort|uniq|awk '{print "- [ ] "$1}'`

- [ ] pytorch_lightning/accelerators (#9558)
- [ ] pytorch_lightning/callbacks
- [ ] pytorch_lightning/callbacks/progress
- [ ] pytorch_lightning/core
- [ ] pytorch_lightning/core/mixins
- [ ] pytorch_lightning/loggers
- [ ] pytorch_lightning/loops
- [ ] pytorch_lightning/loops/dataloader
- [ ] pytorch_lightning/loops/epoch
- [ ] pytorch_lightning/loops/optimization
- [ ] pytorch_lightning/overrides
- [ ] pytorch_lightning/plugins
- [ ] pytorch_lightning/plugins/environments
- [ ] pytorch_lightning/plugins/io
- [ ] pytorch_lightning/plugins/precision
- [ ] pytorch_lightning/plugins/training_type
- [ ] pytorch_lightning/profiler
- [ ] pytorch_lightning/trainer
- [ ] pytorch_lightning/trainer/connectors
- [ ] pytorch_lightning/trainer/connectors/logger_connector
- [ ] pytorch_lightning/tuner
- [ ] pytorch_lightning/utilities

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Part of #5434.

### TODO
- [x] base.py
- [x] early_stopping.py
- [x] finetuning.py
- [x] gpu_stats_monitor.py
- [x] gradient_accumulation_scheduler.py
- [x] \_\_init__.py
- [x] lambda_function.py
- [x] lr_monitor.py
- [x] model_checkpoint.py
- [x] progress.py
- [x] pruning.py

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
You may want to use the issue as a tracker of all the lines and do it on separate PRs. This also allows external contributors

File: pytorch_lightning/callbacks/pruning.py [hunk 1]
```diff
@@ -128,6 +128,14 @@ def __init__(
...
             pruning_norm: if you are using ln_structured you need to specify norm.
 
+        Raises:
+            MisconfigurationException:
+                If ``parameter_names`` is neither ``"weight"`` nor ``"bias"``,
+                if the provided ``pruning_fn`` is not supported,
+                if ``pruning_dim`` is not provided when ``"unstructured"``,
+                if ``pruning_norm`` is not provided when ``"ln_structured"``,
+                if ``pruning_fn`` is neither ``str`` nor :class:`torch.nn.utils.prune.BasePruningMethod`, or
+                if ``amount`` is none of ``int``, ``float`` and ``Callable``.
```
Comment (akihironitta) [inline pytorch_lightning/callbacks/pruning.py:138]: Is this too noisy?

tmp

Comment (Borda) [reply pytorch_lightning/callbacks/pruning.py:138]: yes, can we make it as list?

Comment (akihironitta) [reply pytorch_lightning/callbacks/pruning.py:138]: I actually tried to do so like the lines below, but the output doesn't change to the list view for some reason. Do you have any idea?

https://github.com/PyTorchLightning/pytorch-lightning/blob/84ae9f1ed2b9274783ff93f29e7164ae2e8e6751/pytorch_lightning/callbacks/pruning.py#L113-L117

tmp

Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:138]: You might want to hold on this until #5825 lands

File: pytorch_lightning/callbacks/finetuning.py [hunk 2]
```diff
@@ -318,6 +318,11 @@ def __init__(
         self.verbose = verbose
 
     def on_fit_start(self, trainer, pl_module):
+        """
+        Raises:
+            MisconfigurationException:
+                If LightningModule has no nn.Module `backbone` attribute.
+        """
         if hasattr(pl_module, "backbone") and \
            (isinstance(pl_module.backbone, Module) or isinstance(pl_module.backbone, Sequential)):
```
Comment (Borda) [inline pytorch_lightning/callbacks/finetuning.py:327]:
```suggestion
        if hasattr(pl_module, "backbone") and 
           isinstance(pl_module.backbone, (Module, Sequential)):
```

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 3]
```diff
@@ -115,6 +115,14 @@ class ModelCheckpoint(Callback):
         For example, you can change the default last checkpoint name by doing
         ``checkpoint_callback.CHECKPOINT_NAME_LAST = "{epoch}-last"``
 
+    Raises:
+        MisconfigurationException:
+            If ``save_top_k`` is neither ``None`` nor more than or equal to ``-1``,
+            if ``monitor`` is ``None`` and ``save_top_k`` is none of ``None``, ``-1``, and ``0``, or
+            if ``mode`` is none of ``"min"``, ``"max"``, and ``"auto"``.
```
Comment (Borda) [inline pytorch_lightning/callbacks/model_checkpoint.py:122]: probably list also here

File: pytorch_lightning/callbacks/pruning.py (hunk 1 above)
Comment (Borda) [reply pytorch_lightning/callbacks/pruning.py:138]: ok, lest postponed the Prune, but finish all others...

File: pytorch_lightning/callbacks/pruning.py [hunk 4]
```diff
@@ -332,6 +340,11 @@ def sanitize_parameters_to_prune(
         """
         This function is responsible to check provided `parameters_to_prune` and `parameters`.
         If parameters_to_prune is None, parameters_to_prune will be generated from all parameters of the model.
+
+        Raises:
+            MisconfigurationException:
+                If ``parameters_to_prune`` doesn't exist in the model, or
+                if ``parameters_to_prune`` is neither a list of tuple nor ``None``.
```
Comment (Borda) [inline pytorch_lightning/callbacks/pruning.py:347]: list?

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/callbacks/pruning.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/callbacks/pruning.py:138]: It landed already, so no problem 🙂

File: pytorch_lightning/callbacks/finetuning.py [hunk 5]
```diff
@@ -318,8 +318,12 @@ def __init__(
         self.verbose = verbose
 
     def on_fit_start(self, trainer, pl_module):
-        if hasattr(pl_module, "backbone") and \
-           (isinstance(pl_module.backbone, Module) or isinstance(pl_module.backbone, Sequential)):
```
Comment (akihironitta) [inline pytorch_lightning/callbacks/finetuning.py:322]: `Sequential` is subclass of `Module`, so we don't need this check:
```python
isinstance(pl_module.backbone, Sequential)
```

**Comment by akihironitta** [comment]
Would you mind reviewing this PR again?

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]

**Review by tchaton — COMMENTED** [review]
Great addition !
